api/dataset.go

```diff
@@ -3,37 +3,276 @@ package api
+				logData: logData,
+			},
 		)
 		return
 	}
 
 	api.respond.JSON(ctx, w, http.StatusOK, response)
 }
 
-func (api *API) getDefaultDatasetJSON(ctx context.Context, r *http.Request) (*getDatasetJsonObservationsResponse, error) {
-	return &getDatasetJsonObservationsResponse{}, nil
+func (api *API) getGeographyDatsetJSON(ctx context.Context, r *http.Request) (*getDatasetJSONResponse, error) {
+	return &getDatasetJSONResponse{}, nil
+}
+
+func (api *API) toGetDatasetJsonResponse(params *datasetParams, query *cantabular.StaticDatasetQuery) (*getDatasetJSONResponse, error) {
+	var dimensions []DatasetJSONDimension
+
+	for _, dimension := range query.Dataset.Table.Dimensions {
+		if _, ok := params.options[dimension.Variable.Name]; !ok {
+			return nil, errors.New("dimension mismatch")
+		}
+
+		var options []model.Link
+
+		for _, option := range dimension.Categories {
+			if _, ok := params.options[dimension.Variable.Name][option.Label]; !ok {
+				return nil, errors.New("option mismatch")
+			}
+
+			option := model.Link{
+				HREF: params.options[dimension.Variable.Name][option.Label].Links.Code.URL,
+				ID:   option.Label,
+			}
+
+			options = append(options, option)
+		}
+
+		dimensions = append(dimensions, DatasetJSONDimension{
+			DimensionName: dimension.Variable.Name,
+			Options:       options,
+		})
+	}
+
+	datasetLinks := DatasetJSONLinks{
+		DatasetMetadata: model.Link{
+			HREF: params.metadataLink.URL,
+			ID:   params.metadataLink.ID,
+		},
+		Self: model.Link{
+			HREF: params.datasetLink.URL,
+			ID:   params.datasetLink.ID,
+		},
+		Version: model.Link{
+			HREF: params.versionLink.URL,
+			ID:   params.versionLink.ID,
+		},
+	}
+
+	getDatasetJsonResponse := getDatasetJSONResponse{
+		Dimensions:        dimensions,
+		Links:             datasetLinks,
+		Observations:      query.Dataset.Table.Values,
+		TotalObservations: len(query.Dataset.Table.Values),
+	}
+
+	return &getDatasetJsonResponse, nil
+}
+
+func (api *API) getDatasetParams(ctx context.Context, r *http.Request) (*datasetParams, error) {
+	params := &datasetParams{
+		id:      chi.URLParam(r, "dataset_id"),
+		edition: chi.URLParam(r, "edition"),
+		version: chi.URLParam(r, "version"),
+		options: make(optionsMap),
+	}
+
+	if params.id == "" {
+		return nil, errors.New("invalid dataset id")
+	}
+
+	if params.edition == "" {
+		return nil, errors.New("invalid edition")
+	}
+
+	if params.version == "" {
+		return nil, errors.New("invalid version")
+	}
+
+	datasetItem, err := api.datasets.GetDatasetCurrentAndNext(ctx, "", "", "", params.id)
+	if err != nil {
+		return nil, err
+	}
+
+	params.datasetLink = datasetItem.Links.Self
+	params.basedOn = datasetItem.IsBasedOn.ID
+
+	if datasetItem.DatasetDetails.Type != "cantabular_flexible_table" {
+		return nil, errors.New("invalid dataset type")
+	}
+
+	versionItem, err := api.datasets.GetVersion(ctx, "", "", "", "", params.id, params.edition, params.version)
+	if err != nil {
+		return nil, err
```

Comment:
Can we make sure we're wrapping errors each time we're returning? Just a brief description of what's failed will do, it leads to much more useful error messages e.g.
```
return nil, errors.Wrap(err, "failed to get version")
```